src/prefect/records/base.py

```diff
@@ -0,0 +1,152 @@
+import os
+import socket
+import threading
+from contextlib import contextmanager
+from dataclasses import dataclass
+from typing import Optional
+
+from prefect.results import BaseResult
+
+
+class RecordStore:
+    def read(self, key: str):
+        raise NotImplementedError
+
+    def write(self, key: str, value: dict):
+        raise NotImplementedError
```

Comment:
What are the return types on these?

Reply:
The signatures of the `MemoryRecordStore` and the `ResultFactoryStore` don't align right now, so I'm leaving the typing on the existing methods as is for now. I plan to update the typing/signature for `RecordStore` and `ResultFactoryStore` when I get `Transaction` working with `MemoryRecordStore` in the next PR.

Reply:
Sounds great!